### ISSUE TITLE
Crashes while failing to encode an inner let rec

### ISSUE THREAD

**Issue by nikswamy** [FStarLang/FStar#1362]
This program, with `fstar.exe --debug Test --debug_level SMTQuery --trace_error`

```
module Test
let f (n:nat) : Tot nat =
  let k = 
    let rec g (m:nat) : Tot nat =
      if m = 0 then 0
      else g (m - 1)
    in 
    g n 
  in
  k + 1
```

results in a 

```
Encoding query formula: forall (n:Prims.nat).
  (forall (m:Prims.nat) (g:(m:Prims.nat{ m << m } -> Prims.Tot Prims.nat)).
      {:pattern
       could not prove post-condition
      }
      (*Subtyping check failed; expected type (*?u1548*) _ n m g; got type Prims.nat*) Prims.l_True /\
      (forall (y:Prims.nat).
          y == m ==>
          (*Subtyping check failed; expected type (*?u1548*) _ n m g; got type Prims.int*)
          Prims.l_True /\
          (forall (y:Prims.int).
              y == 0 ==>
              (forall (y:Prims.bool).
                  y == (m = 0) ==>
                  y == (m = 0) ==>
                  (forall (k:Prims.pure_post Prims.nat).
                      (forall (x:Prims.nat).
                          {:pattern
                           Prims.guard_free (k x)
                          }
                          k x <==> Prims.l_True) ==>
                      m = 0 == true ==> (forall (y:Prims.int). y == 0 ==> k y) /\
                      ~(m = 0 == true) ==>
                      (forall (b:Prims.bool).
                          m = 0 == b ==>
                          (*Subtyping check failed; expected type (m:Prims.nat{ m << m }); got type Prims.int*)
                          Prims.b2t (m - 1 >= 0) /\ m - 1 << m /\
                          (forall (y:(m:Prims.nat{ m << m })).
                              y == m - 1 ==> (forall (y:Prims.nat). y == g y ==> k y)))))))) /\
  (*could not prove post-condition*)
  (*Subtyping check failed; expected type Prims.nat; got type Prims.int*)
  Prims.b2t ((if n = 0
        then 0
        else
          (let rec g m =
              ((if m = 0 <: Prims.bool then 0 else g (m - 1)) <: Prims.nat) <: Prims.Tot Prims.nat
            in
            g) (n - 1)) +
      1 >=
      0)
.\test.fst(4,4-8,7): (Info) Non-top-level recursive functions, and their enclosings let bindings (including the top-level let) are not yet fully encoded to the SMT solver; you may not be able to prove some facts
Unexpected error
Exception of type 'FStar.SMTEncoding.Encode+Inner_let_rec' was thrown.
   at FStar.SMTEncoding.Encode.encode_term(syntax`1 t, env_t env)
   at FStar.SMTEncoding.Encode.encode_partial_app@938(env_t env, syntax`1 head, FSharpList`1 args_e, FSharpList`1 decls, FSharpList`1 args, FSharpOption`1 ht_opt)
   at FStar.SMTEncoding.Encode.encode_term(syntax`1 t, env_t env)
   at FStar.SMTEncoding.Encode.encode_match@1132-1.Invoke(Tuple`2 tupledArg)
```

Whereas, rewriting it like so, does not:

```
module Test
let f (n:nat) : Tot nat =
  let rec g (m:nat) : Tot nat =
    if m = 0 then 0
    else g (m - 1)
  in 
  let k = g n  in
  k + 1
```

I'm trying to fix it now. But, heads up @aseemr  ... you may find this interesting. I think it's related to our revised version of Tot term inlining and this comment:

https://github.com/FStarLang/FStar/blob/a1bbe57ff3382104d6feab602b3039d9f7abb891/src/typechecker/FStar.TypeChecker.Util.fs#L653

**Comment by mtzguido** [comment]
Seems to work now, should this have been closed?

**Comment by nikswamy** [comment]
Yes, this was fixed. But, I think there should be a better fix once we solve #851
